probe/awsecs/client.go

```diff
@@ -106,45 +216,173 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
+			continue
+		}
+		serviceName := serviceNameRaw.(string)
+		service := serviceRaw.(EcsService)
+		for _, deployment := range service.DeploymentIDs {
+			deploymentMap[deployment] = serviceName
+		}
+	}
+	log.Debugf("Mapped %d deployments from %d services", len(deploymentMap), c.serviceCache.Len())
+
+	results := map[string]string{}
+	unmatched := []string{}
+	for _, taskARN := range taskARNs {
+		taskRaw, err := c.taskCache.Get(taskARN)
+		if err != nil {
+			// this can happen if we have a failure while describing tasks, just pretend the task doesn't exist
+			continue
+		}
+		task := taskRaw.(EcsTask)
+		if !strings.HasPrefix(task.StartedBy, servicePrefix) {
+			// task was not started by a service
+			continue
+		}
+		if serviceName, ok := deploymentMap[task.StartedBy]; ok {
+			results[taskARN] = serviceName
+		} else {
+			unmatched = append(unmatched, taskARN)
+		}
+	}
+
+	log.Debugf("Matched %d from %d tasks, %d unmatched", len(results), len(taskARNs), len(unmatched))
+	return results, unmatched
+}
+
+func (c ecsClientImpl) ensureTasks(taskARNs []string) {
+	tasksToFetch := []string{}
+	for _, taskARN := range taskARNs {
+		if _, err := c.taskCache.Get(taskARN); err != nil {
+			tasksToFetch = append(tasksToFetch, taskARN)
+		}
+	}
+	if len(tasksToFetch) > 0 {
+		// This might not fully succeed, but we only try once and ignore any further missing tasks.
+		c.getTasks(tasksToFetch)
+	}
+}
+
+func (c ecsClientImpl) refreshServices(taskServiceMap map[string]string) map[string]bool {
+	toDescribe, done := c.describeServices()
+	servicesRefreshed := map[string]bool{}
+	for _, serviceName := range taskServiceMap {
+		if servicesRefreshed[serviceName] {
+			continue
+		}
+		toDescribe <- serviceName
+		servicesRefreshed[serviceName] = true
 	}
-	return results, nil
+	close(toDescribe)
+	<-done
+	return servicesRefreshed
 }
 
-// returns a map from task ARNs to service names
-func (c ecsClient) getInfo(taskArns []string) (ecsInfo, error) {
-	servicesChan := make(chan map[string]*ecs.Service)
+func (c ecsClientImpl) describeAllServices(servicesRefreshed map[string]bool) {
+	serviceNamesChan := c.listServices()
+	toDescribe, done := c.describeServices()
 	go func() {
-		servicesChan <- c.getServices()
+		for serviceName := range serviceNamesChan {
+			if !servicesRefreshed[serviceName] {
+				toDescribe <- serviceName
+				servicesRefreshed[serviceName] = true
+			}
+		}
+		close(toDescribe)
 	}()
+	<-done
+}
 
-	// do these two fetches in parallel
-	tasks, err := c.getTasks(taskArns)
-	services := <-servicesChan
+func (c ecsClientImpl) makeECSInfo(taskARNs []string, taskServiceMap map[string]string) EcsInfo {
+	// The maps to return are the referenced subsets of the full caches
+	tasks := map[string]EcsTask{}
+	for _, taskARN := range taskARNs {
+		// It's possible that tasks could still be missing from the cache if describe tasks failed.
+		// We'll just pretend they don't exist.
+		if taskRaw, err := c.taskCache.Get(taskARN); err == nil {
```

Comment:
Worth factoring it out in a function (this casting seems to be used at at least a couple of places)? (same for services)

Maybe add a cache interface?